### PR TITLE
PLANNER-1818 Fix ConstraintProvider for Machine Reassignment example

### PR DESCRIPTION
**Pull request by rsynek**

### CODE REVIEW

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 1]
```diff
@@ -53,31 +62,28 @@
...
+                .join(MrProcessAssignment.class,
+                      equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
+                )
+                .groupBy((machineCapacity, processAssignment) -> machineCapacity, sumLong(
+                        (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource()))
+                )
+                .filter(((machineCapacity, usage) -> machineCapacity.getMaximumCapacity() < usage))
+                .penalizeLong(MrConstraints.MAXIMUM_CAPACITY,
+                              HardSoftLongScore.ofHard(1L),
+                              (machineCapacity, usage) -> machineCapacity.getMaximumCapacity() - usage);
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: If the usage is higher than the capacity, doesn't `capacity - usage` return a negative number? But we have a fail fast for that.
What am I missing?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 2]
```diff
@@ -53,31 +62,28 @@
...
-//                .penalizeLong(MrConstraints.MAXIMUM_CAPACITY,
-//                        HardSoftLongScore.ofHard(1L),
-//                        (machineCapacity, usage) -> machineCapacity.getMaximumCapacity() - usage);
+        return factory.from(MrMachineCapacity.class)
+                .join(MrProcessAssignment.class,
+                      equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
+                )
+                .groupBy((machineCapacity, processAssignment) -> machineCapacity, sumLong(
+                        (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource()))
+                )
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: One experiment to make this go faster (not to be done as part of this PR!),
is to something like this:

``
from(MrProcessAssignment) // Look mom, no resource selected!
.groupBy(MrProcessAssignment::getMachine,
    sum(MrProcessAssignment::getUsages, some lamdba's)) // MrProcessAssignment::getUsages returns an array of usages
```

Now that would screw up our constraint matches count, etc but the score would end up the same. Regardless if we'd want to do it, it's an interesting performance experiment.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 3]
```diff
@@ -139,42 +145,40 @@ private Constraint transientUsage(ConstraintFactory factory) {
...
+                .penalizeLong(MrConstraints.BALANCE_COST, HardSoftLongScore.ofSoft(1L), this::balanceCost);
+    }
+
+    private long balanceCost(MrBalancePenalty penalty, MrMachine machine, long originalUsage, long targetUsage) {
+        long originalAvailability =
+                machine.getMachineCapacity(penalty.getOriginResource()).getMaximumCapacity() - originalUsage;
+        long targetAvailability =
+                machine.getMachineCapacity(penalty.getTargetResource()).getMaximumCapacity() - targetUsage;
+        long balanceCost = penalty.getMultiplicand() * originalAvailability - targetAvailability;
+        return balanceCost < 0 ? 0 : balanceCost;
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: `Math.max(0, balanceCost)`?

Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Not sure if that's more readable. Or faster.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 4]
```diff
@@ -139,42 +145,40 @@ private Constraint transientUsage(ConstraintFactory factory) {
...
-//                .filter(((machineCapacity, usage) -> machineCapacity.getSafetyCapacity() < usage))
-//                .penalizeLong(MrConstraints.LOAD_COST,
-//                        HardSoftLongScore.ofSoft(1L),
-//                        (machineCapacity, usage) -> machineCapacity.getSafetyCapacity() - usage);
+        return factory.from(MrMachineCapacity.class)
+                .join(MrProcessAssignment.class,
+                      equal(MrMachineCapacity::getMachine, MrProcessAssignment::getMachine)
+                )
+                .groupBy((machineCapacity, processAssignment) -> machineCapacity, sumLong(
+                        (machineCapacity, processAssignment) -> processAssignment.getUsage(machineCapacity.getResource()))
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Any chance the `sumLong()` could come on a new line, like you do in other constraints?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java (hunk 3 above)
Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Although performance was not my concern here, although `Math.max()` is intrinsic in Hotspot.
Mainly I'm interested in one less condition we have to test. :-)

Anyway... fine to merge with or without this.

Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Same here, fine to merge with or without this, too. You're owner, your call :)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 5]
```diff
@@ -184,23 +188,19 @@ private Constraint processMoveCost(ConstraintFactory factory) {
         return factory.from(MrProcessAssignment.class)
                 .filter(MrProcessAssignment::isMoved)
                 .penalizeLong(MrConstraints.PROCESS_MOVE_COST,
-                        HardSoftLongScore.ofSoft(1L),
-                        MrProcessAssignment::getProcessMoveCost);
+                              HardSoftLongScore.ofSoft(1L),
+                              MrProcessAssignment::getProcessMoveCost);
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: We don't align on columns.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 6]
```diff
@@ -210,8 +210,7 @@ private Constraint machineMoveCost(ConstraintFactory factory) {
         return factory.from(MrProcessAssignment.class)
                 .filter(MrProcessAssignment::isMoved)
                 .penalizeLong(MrConstraints.MACHINE_MOVE_COST,
-                        HardSoftLongScore.ofSoft(1L),
-                        MrProcessAssignment::getMachineMoveCost);
+                              HardSoftLongScore.ofSoft(1L),
+                              MrProcessAssignment::getMachineMoveCost);
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Dtto. on aligning.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java [hunk 7]
```diff
@@ -53,31 +62,28 @@
...
      */
     private Constraint serviceConflict(ConstraintFactory factory) {
         return factory.fromUniquePair(MrProcessAssignment.class,
-                equal(MrProcessAssignment::getMachine),
-                equal(MrProcessAssignment::getService)
+                                      equal(MrProcessAssignment::getMachine),
+                                      equal(MrProcessAssignment::getService)
         ).penalize(MrConstraints.SERVICE_CONFLICT,
-                HardSoftLongScore.ofHard(1L));
+                   HardSoftLongScore.ofHard(1L));
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: FYI All instances of `HardSoftLongScore.ofHard(1L)` may be replaced by `HardSoftLongScore.ONE_HARD`,

Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: (Similarly for `ONE_SOFT`.)

File: optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/machinereassignment/solver/score/MachineReassignmentConstraintProvider.java]: Good catch; there was no machine breaking this constraint. Once I removed the filter, the fast fast worked.

**Pull request merged**